File: libraries/listing/src/main/kotlin/orwir/starrit/listing/feed/Post.kt
package orwir.starrit.listing.feed

import android.content.res.Resources
import android.net.Uri
import androidx.core.net.toUri
import orwir.starrit.core.extension.toHtml
import orwir.starrit.core.util.createHashCode
import orwir.starrit.core.util.prettyDate
import orwir.starrit.core.util.squeeze
import orwir.starrit.listing.R
import orwir.starrit.listing.model.Image
import orwir.starrit.listing.model.Submission
import orwir.starrit.listing.model.Subreddit
import orwir.starrit.listing.util.videoOrNull
import java.io.Serializable

sealed class Post(submission: Submission, res: Resources) : Serializable {
    val id: String = submission.id
    val subreddit: Subreddit = submission.subreddit
    val author: String = submission.author
    val created: String = prettyDate(submission.created * 1000, res)
    val title: String = submission.title
    val nsfw: Boolean = submission.nsfw
    val spoiler: Boolean = submission.spoiler
    val comments: String = submission.commentsCount.squeeze()
    val score: String = prettyScore(submission, res)
    val domain: String = submission.domain
    val selfDomain: Boolean = domain.startsWith("self.")
    val contentUrl: String = submission.url
    val postUrl: String = submission.permalink
    val imagePreview: Image = submission.preview
        ?.images
        ?.firstOrNull()
        ?.resolutions
        ?.firstOrNull()
        ?: Image("", 0, 0)
    val imageSource: Image = submission.preview
        ?.images
        ?.firstOrNull()
        ?.source
        ?: Image("", 0, 0)
    val imageBlurred: Image = submission
        .preview
        ?.images
        ?.firstOrNull()
        ?.variants
        ?.nsfw
        ?.resolutions
        ?.firstOrNull()
        ?: Image("", 0, 0)

    override fun equals(other: Any?) = other is Post
            && other::class == this::class
            && id == other.id
            && subreddit == other.subreddit
            && author == other.author
            && created == other.created
            && title == other.title
            && nsfw == other.nsfw
            && spoiler == other.spoiler
            && comments == other.comments
            && score == other.score
            && domain == other.domain
            && contentUrl == other.contentUrl
            && postUrl == other.postUrl
            && imagePreview == other.imagePreview
            && imageSource == other.imageSource
            && imageBlurred == other.imageBlurred

    override fun hashCode() = createHashCode(
        id,
        subreddit,
        author,
        created,
        title,
        nsfw,
        spoiler,
        comments,
        score,
        domain,
        contentUrl,
        postUrl,
        imagePreview,
        imageSource,
        imageBlurred
    )

    private fun prettyScore(submission: Submission, res: Resources) =
        if (submission.hideScore) {
            res.getString(R.string.post_score_hidden)
        } else {
            submission.score.squeeze()
        }
}

class LinkPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val link: Uri = contentUrl.toUri()
    val displayLink: String = link.authority ?: submission.url
}

class ImagePost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources)

class GifPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val gif: String = submission.url
}

class TextPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val text: CharSequence = (submission.selftextHtml ?: submission.selftext ?: "").toHtml()
}

class VideoPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val video: String = submission.videoOrNull() ?: ""
}

File: lib/model/post.dart
import 'package:flutter/foundation.dart';
import 'package:starrit/model/author.dart';
import 'package:starrit/model/image.dart';
import 'package:starrit/model/subreddit.dart';
import 'package:starrit/util/json.dart';
import 'package:starrit/util/content.dart';

@immutable
class Post {
  final String id;
  final Subreddit subreddit;
  final Author author;
  final DateTime created;
  final String title;
  final bool nsfw;
  final bool spoiler;
  final int comments;
  final int score;
  final String domain;
  final bool selfDomain;
  final String postUrl;
  final String contentUrl;
  final ContentType type;
  final ImageData imagePreview;
  final ImageData imageSource;
  final ImageData imageBlurred;
  final String text;
  final String gif;
  final String video;

  Post({
    @required this.id,
    @required this.subreddit,
    @required this.author,
    @required this.created,
    @required this.title,
    @required this.nsfw,
    @required this.spoiler,
    @required this.comments,
    @required this.score,
    @required this.domain,
    @required this.selfDomain,
    @required this.postUrl,
    @required this.contentUrl,
    @required this.type,
    @required this.imagePreview,
    @required this.imageSource,
    @required this.imageBlurred,
    @required this.text,
    @required this.gif,
    @required this.video,
  });

  Post.fromJson(Map<String, dynamic> submission)
      : this(
          id: submission.get('name'),
          subreddit: Subreddit.fromJson(submission),
          author: Author.fromJson(submission),
          created: DateTime.fromMillisecondsSinceEpoch(
            submission.get<double>('created_utc').toInt() * 1000,
            isUtc: true,
          ),
          title: submission.get('title'),
          nsfw: submission.get('nsfw') ?? false,
          spoiler: submission.get('spoiler') ?? false,
          comments: submission.get('num_comments'),
          score: submission.get('score'),
          domain: submission.domain,
          selfDomain: submission.isSelfDomain,
          postUrl: submission.get('permalink'),
          contentUrl: submission.url,
          type: submission.contentType,
          imagePreview: submission.image('preview.images[0].resolutions[0]'),
          imageSource: submission.image('preview.images[0].source'),
          imageBlurred: submission.image(
            'preview.images[0].variants.nsfw.resolutions[0]',
          ),
          text: submission.text,
          gif: submission.gif,
          video: submission.video,
        );

  @override
  int get hashCode {
    return id.hashCode ^
        subreddit.hashCode ^
        author.hashCode ^
        created.hashCode ^
        title.hashCode ^
        nsfw.hashCode ^
        spoiler.hashCode ^
        comments.hashCode ^
        domain.hashCode ^
        postUrl.hashCode ^
        contentUrl.hashCode ^
        (imagePreview?.hashCode ?? 0) ^
        (imageSource?.hashCode ?? 0) ^
        (imageBlurred?.hashCode ?? 0);
  }

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is Post &&
            runtimeType == other.runtimeType &&
            id == other.id &&
            subreddit == other.subreddit &&
            author == other.author &&
            created == other.created &&
            title == other.title &&
            nsfw == other.nsfw &&
            spoiler == other.spoiler &&
            comments == other.comments &&
            domain == other.domain &&
            postUrl == other.postUrl &&
            contentUrl == other.contentUrl &&
            imagePreview == other.imagePreview &&
            imageSource == other.imageSource &&
            imageBlurred == other.imageBlurred;
  }
}
